src/_guides/language/language-tour.md

```diff
@@ -1723,32 +1723,38 @@ assert(foo() == null);
+| bitwise XOR                             | `^`                                                                                                                                                                                               | Left          |
+| bitwise OR                              | `|`                                                                                                                                                                                               | Left          |
+| relational&nbsp;and&nbsp;type&nbsp;test | `>=`    `>`    `<=`    `<`    `as`    `is`    `is!`                                                                                                                                               | None          |
+| equality                                | `==`    `!=`                                                                                                                                                                                      | None          |
+| logical AND                             | `&&`                                                                                                                                                                                              | Left          |
+| logical OR                              | `||`                                                                                                                                                                                              | Left          |
+| if null                                 | `??`                                                                                                                                                                                              | Left          |
+| conditional                             | <code><em>expr1</em> ? <em>expr2</em> : <em>expr3</em></code>                                                                                                                                     | Right         |
+| cascade                                 | `..` &nbsp;&nbsp; `?..`                                                                                                                                                                           | Right         |
+| assignment                              | `=`    `*=`    `/=`    `+=`    `-=`    `&=`    `^=`    <em>etc.</em>                                                                                                                              | Right         |
+
 {:.table .table-striped}
 
 {{site.alert.warning}}
-  Operator precedence is an approximation of the behavior of a Dart parser.
+  The operator precedence outlined here
+  is only an approximation of the precedence and associativity
+  given implicitly by Dart's grammar
+  and should only be used as a helpful guide.
   For definitive answers, consult the grammar in the
   [Dart language specification][].
```

Comment:
**question:** Why does this only approximate precedence? Is there a difficulty in providing the specific precedence? What is that difficulty?

**suggestion:** Tighten the language and use active voice.
```suggestion
  The operator precedence that the previous table outline provides 
  a rough guide. You can find the precedence and associativity for Dart
  in the [Dart language specification][].
```

Reply:
It's not that you can find any authoritative precedence and associativity in the language specification as they aren't really defined (or able to be exactly), but rather you need to look at the whole grammar to understand the relationships between operators.

I've took some of your changes and Erik's, while also being clear about it only be used as a guide (which was the source of confusion that caused the introduction of the note a long time ago).

Reply:
I could comment a bit more on why operator precedence is not a precise concept.

If you settle on a specific class of parsers (say, table driven LALR(1) parsers) then you _can_ have a technically precise notion of precedence. In case of an ambiguity, the parser will then use the assigned precedence values to make the decision about whether to shift or reduce, and which rule to reduce. This makes sense if we have a grammar where a lot of different expressions use the same nonterminal:

```
...
expr ::= expr ('+' | '-') expr @precedence(5)
expr ::= expr ('*' | '/') expr @precedence(6)
...
```

Precedence has a precise meaning in this context, and it is a tool outside the underlying context free grammar formalism (it is defined in terms of the chosen parser algorithm, it's not a property of the grammar &mdash; the grammar is simply ambiguous).

If we are using a different approach to parsing (for instance, Dart has always used an enhanced recursive descent parser, which is more like ANTLR) then there is no reason to assume that there exists any assignment of precedence to any set of operators that will describe the behavior of the parser correctly. In other words, the whole concept of precedence doesn't necessarily fit that approach to parsing. When using that approach, the ambiguities simply do not arise, because we use different nonterminals to establish a similar property:

```
...
additiveExpression ::= multiplicativeExpression (('+' | '-') multiplicativeExpression)*
multiplicativeExpression ::= someOtherExpression (('*' | '/') someOtherExpression)*
...
```

Looking at the grammar you can see that an additive expression can consist of a sequence of multiplicative expressions, separated by '+' or '-' (and not the other way around), and that makes it tempting to say that '*' has a higher precedence than '+'. (Of course, there would be more rules about `additiveExpression`, because it can also derive other things than a sum of products, but this is the core part with respect to precedence.)

However, the source of truth is the grammar, and if we settle on a particular set of precedence values, it takes a non-trivial proof to verify that this precedence table describes the derivable programs with that grammar precisely (and maybe that proof doesn't exist, because it just isn't true).

Reply:
@parlough : I'd still recommend the tighter language.

@eernstg : Thanks for the explanation!